test/performance/odp_ipsecfwd.c

```diff
@@ -0,0 +1,1307 @@
+	       "                          1: direct output\n"
+	       "  -s, --sa            SA configuration file. Individual SA configuration is\n"
+	       "                      expected to be within a single line, values whitespace\n"
+	       "                      separated:\n"
+	       "\n"
+	       "                          <line in file> Dir SPI SrcIPv4 DstIPv4 CipherAlgoIdx"
+	       " CipherKey CipherKeyExtra AuthAlgIdx AuthKey AuthKeyExtra ICVLen\n"
+	       "\n"
+	       "                      With combined algorithms, authentication data is ignored.\n"
+	       "                      Traffic is mapped to SAs based on UDP port: the port is\n"
+	       "                      used as the SPI. Non-zero Dir value declares an outbound\n"
+	       "                      SA whereas zero Dir value declares an inbound SA.\n"
+	       "\n"
+	       "                      Supported cipher and authentication algorithms:\n",
+	       PROG_NAME, PROG_NAME, MIN(pool_capa.pkt.max_num, PKT_CNT),
+	       MIN(pool_capa.pkt.max_len, PKT_SIZE), WORKER_CNT);
+	print_supported_algos(&ipsec_capa);
+	printf("  -f, --fwd_table     Forwarding configuration file. Individual forwarding\n"
+	       "                      configuration is expected to be within a single line,\n"
+	       "                      values whitespace separated:\n"
+	       "\n"
+	       "                          <line in file> DstIPv4 SubNetMask NetIf DstMac\n"
+	       "\n"
+	       "                      DstIPv4 and SubNetMask define a matchable prefix and NetIf\n"
+	       "                      and DstMac define the outgoing interface and destination\n"
+	       "                      MAC address for a match. NetIf should be one of the\n"
+	       "                      interfaces passed with \"--interfaces\" option\n"
+	       "  -h, --help          This help.\n"
+	       "\n");
+}
+
+static odp_bool_t setup_ipsec(prog_config_t *config)
+{
+	odp_queue_param_t q_param;
+	odp_ipsec_config_t ipsec_config;
+
+	odp_queue_param_init(&q_param);
+	q_param.type = ODP_QUEUE_TYPE_SCHED;
+	q_param.sched.prio = odp_schedule_default_prio();
+	q_param.sched.sync = ODP_SCHED_SYNC_PARALLEL;
+	q_param.sched.group = ODP_SCHED_GROUP_ALL;
+	config->compl_q = odp_queue_create(PROG_NAME, &q_param);
+
+	if (config->compl_q == ODP_QUEUE_INVALID) {
+		ODPH_ERR("Error creating IPsec completion queue\n");
+		return false;
+	}
+
+	odp_ipsec_config_init(&ipsec_config);
+	ipsec_config.inbound_mode = ODP_IPSEC_OP_MODE_ASYNC;
+	ipsec_config.outbound_mode = ODP_IPSEC_OP_MODE_ASYNC;
+	ipsec_config.inbound.default_queue = config->compl_q;
+	/* For tunnel to tunnel, we need to parse up to this to check the UDP port for SA. */
+	ipsec_config.inbound.parse_level = ODP_PROTO_LAYER_L4;
+
+	if (odp_ipsec_config(&ipsec_config) < 0) {
+		ODPH_ERR("Error configuring IPsec\n");
+		return false;
+	}
+
+	return true;
+}
+
+static odp_bool_t check_alg_support(int cipher_idx)
```

Comment:
Maybe alg_idx instead of cipher_idx?

I am not sure how useful this support check and the concept of "exposed algorithms" is. Printing the ids in usage is useful, I think, but not allowing to use any algorithm that is not known by this test tool is not very helpful.

Now that this test SW does not know all algorithms, it is annoying that the unknown algorithms are disabled even though they should work just fine. And once all the known algorithms get added in the list, the check does not really serve any purpose.

It would be more useful to prevent usage of algorithms that are not supported by the underlying ODP, but there is no such capability check anywhere. odph_ipsec_alg_check() helper function may be useful in adding such check.

Reply:
Dropped this entirely in v4. The listing created in help should be clear enough what is supported and what is not and if user gets adventurous with the indices it just leads to undefined behavior.

Added also the rest of the algorithms supported by spec.
```
...
  Supported cipher and authentication algorithms:
      Cipher algorithms:
          0: ODP_CIPHER_ALG_NULL (key lengths: 0, 0)
          2: ODP_CIPHER_ALG_3DES_CBC (key lengths: 24)
          4: ODP_CIPHER_ALG_AES_CBC (key lengths: 16, 24, 32)
          5: ODP_CIPHER_ALG_AES_CTR (key lengths: 16, 24, 32)
          9: ODP_CIPHER_ALG_AES_GCM (combined) (key lengths: 16, 24, 32)
          11: ODP_CIPHER_ALG_CHACHA20_POLY1305 (combined) (key lengths: 32)
      Authentication algorithms:
          0: ODP_AUTH_ALG_NULL (key/icv lengths: 0/0, 0/0)
          1: ODP_AUTH_ALG_MD5_HMAC (key/icv lengths: 16/12)
          2: ODP_AUTH_ALG_SHA1_HMAC (key/icv lengths: 20/12)
          4: ODP_AUTH_ALG_SHA256_HMAC (key/icv lengths: 32/16)
          5: ODP_AUTH_ALG_SHA384_HMAC (key/icv lengths: 48/24)
          6: ODP_AUTH_ALG_SHA512_HMAC (key/icv lengths: 64/32)
          7: ODP_AUTH_ALG_AES_GCM (combined) (key/icv lengths: 0/16)
          8: ODP_AUTH_ALG_AES_GMAC (key/icv lengths: 16/16, 24/16, 32/16)
          10: ODP_AUTH_ALG_AES_CMAC (key/icv lengths: 16/12, 24/12, 32/12)
          11: ODP_AUTH_ALG_AES_XCBC_MAC (key/icv lengths: 16/12)
          12: ODP_AUTH_ALG_CHACHA20_POLY1305 (combined) (key/icv lengths: 0/16)
...
```

Reply:
The ODP API is perhaps a bit weird that it reuses the algorithm names defined in the crypto API for IPsec even though IPsec is defined only for the algorithms for which a "use-of-algorithm-x-with-ipsec" spec exists. IOW, some of these do not make sense with IPsec, at least not in any standard-defined way, but it is not easy to tell which unless you already know it.

It would be better to not show the undefined (for IPsec) algorithms. It makes things less confusing and the undefined ones would not be supported by any ODP anyway. These should be removed:

ALG_ENTRY(ODP_CIPHER_ALG_3DES_ECB, CIPHER_TYPE),
ALG_ENTRY(ODP_CIPHER_ALG_AES_CFB128, CIPHER_TYPE),
ALG_ENTRY(ODP_CIPHER_ALG_AES_XTS, CIPHER_TYPE),
ALG_ENTRY(ODP_CIPHER_ALG_KASUMI_F8, CIPHER_TYPE),
ALG_ENTRY(ODP_CIPHER_ALG_SNOW3G_UEA2, CIPHER_TYPE),
ALG_ENTRY(ODP_CIPHER_ALG_AES_EEA2, CIPHER_TYPE),
ALG_ENTRY(ODP_CIPHER_ALG_ZUC_EEA3, CIPHER_TYPE),
ALG_ENTRY(ODP_AUTH_ALG_KASUMI_F9, AUTH_TYPE),
ALG_ENTRY(ODP_AUTH_ALG_SNOW3G_UIA2, AUTH_TYPE),
ALG_ENTRY(ODP_AUTH_ALG_AES_EIA2, AUTH_TYPE),
ALG_ENTRY(ODP_AUTH_ALG_ZUC_EIA3, AUTH_TYPE),
ALG_ENTRY(ODP_AUTH_ALG_MD5, AUTH_TYPE),
ALG_ENTRY(ODP_AUTH_ALG_SHA1, AUTH_TYPE),
ALG_ENTRY(ODP_AUTH_ALG_SHA224, AUTH_TYPE),
ALG_ENTRY(ODP_AUTH_ALG_SHA256, AUTH_TYPE),
ALG_ENTRY(ODP_AUTH_ALG_SHA384, AUTH_TYPE),
ALG_ENTRY(ODP_AUTH_ALG_SHA512, AUTH_TYPE)

Reply:
Now after adding the comment here, I realize that there is a capability check so the undefined ones are not actually displayed to the user. But I would still remove them from the code too.